Enhancement doc to customize default pre-pinned resources on navigation (#1285)
openshift/enhancements: enhancements/console/pinned-resources.md

|           ├── ...
|           ├── pinnedResources
|                   └──
└── ...
```

Each perspective conforms to the following schema :

If the `pinnedResources` list is empty, the default pinned resources will be shown.
If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.

### API Extensions

Extend the existing `operator.openshift.io/v1` / `Console CRD` by adding `spec.customization.perspectives.pinnedResources`

So the example yaml should look like this:

To set default pre-pinned resources

```yaml
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
  ...
spec:
  customization:
    perspectives:
    - id: 'dev'
      ...
      pinnedResources: ['deployments', 'deploymentconfigs']
    - ...
  ...
```

### Risks and Mitigations

N/A

### Drawbacks

N/A

## Design Details

We will provide a configuration UI and a yaml snippet in the
console to provide a list of resources that can be pinned to the navigation.

### Test Plan

- If no pinnedResources are defined, then default pinned resources should be visible to users who do not have their navigation items customized

> - If pinnedResources is an empty list, then default pinned resources should be visible to users who do not have their navigation items customized

Reviewer: where will these default pinned resources be fetched from if pinnedResources is an empty list?

Author: they are contributed by the plugins.
Ex: our dev perspective plugin has the default pins
```
{
    "type": "console.perspective",
    "properties": {
      "id": "dev",
      "name": "%devconsole~Developer%",
      "icon": { "$codeRef": "perspective.icon" },
      "defaultPins": [
        {
          "group": "",
          "version": "v1",
          "kind": "ConfigMap"
        },
        {
          "group": "",
          "version": "v1",
          "kind": "Secret"
        }
      ],
      "landingPageURL": { "$codeRef": "perspective.getLandingPageURL" },
      "importRedirectURL": { "$codeRef": "perspective.getImportRedirectURL" },
      "usePerspectiveDetection": { "$codeRef": "perspective.usePerspectiveDetection" }
    }
  },

Reviewer 2: :warning: @debsmita1 @JoelSpeed As this is not our daily business: We CAN (and is it okay??) to differentiate here between not configured (null) and an empty list?

> - If the pinnedResources list is non-empty, then users who do not have their navigation customized should see the pinned resources defined in the customization

Reviewer: how do we check if the user has done the customization or not?

Author: if the user customize the navigation items it will saved in the user-settings

  - "@spadgett"
api-approvers:
  - "@JoelSpeed"
creation-date: 2022-11-15
last-updated: 2022-11-15
tracking-link:
  - https://issues.redhat.com/browse/ODC-5012
status: implementable
---

# Define pre-pinned resources

## Summary

The Developer perspective on the OpenShift Console provides pinned resources on the navigation which enables the user to go to the resource’s list page with a single click.

Some customers want to provide defaults for those pre-pinned resources for all users.

On the Developer perspective navigation, users can see the default pinned resources if they have not customized these pinned resources. The customers have no way to set these default pinned resources for their users.

This enhancement allows cluster admins to define pre-pinned resources for new users and users who have not customized their navigation items.

![Default pinned resources](default-pinned-resources.png)

## Motivation

Some customers want to provide default pre-pinned resources for all users.

### User Stories

#### Story 1

Customer wants to customize the developer perspective navigation by providing pre-pinned resources to users who do not have their navigation items customized.

> As a cluster admin, I want to provide default pre-pinned resources for all users.

Reviewer 3: Your user stories are missing the ending.

Normally, user stories are of the form "As a <person>, I want to <do something>, so that <I can achieve some goal>", currently the goal part is missing here, you are telling me what they want to do, but not why, and the why is important to help me understand why you are implementing the feature.

### Goals

- Allow cluster admins to define default pre-pinned resources on the dev perspective navigation

### Non-Goals

N/A

## Proposal

We can extend the existing `operator.openshift.io/v1` / `Console CRD`.
We can add `spec.customization.perspectives.pinnedResources` which lists default pre-pinned resources

### Workflow Description

To add default pre-pinned resources:

```text
├── spec
│   ├── customization
|       ├── perspectives
|           ├── id
|           ├── ...
|           ├── pinnedResources
|               └── group
                └── version
                └── kind
└── ...
```

Each perspective conforms to the following schema :

- `pinnedResources: authV1.GroupVersionKind[]` , defines a list of default pinned resources
- `group: string` , defines the group of the resource type.
- `version: string` , defines the version of the resource type.

> - `kind: string` , defines the kind of the resource type.

Reviewer 3: What happens if an invalid combination of these is provided by the user?

Reviewer 2: @debsmita1 @JoelSpeed are you find with `group`, `version`, `kind`, or isn't `apiVersion` (group/version) and `kind` more common?

Like:

```yaml
apiVersion: v1
kind: ConfigMap
---
apiVersion: apps/v1
kind: Deployment
---
apiVersion: apps.openshift.io/v1
kind: DeploymentConfig
```

> If the `pinnedResources` list is empty, the default pinned resources will be shown.

Reviewer 3: I'd like to see documentation on the field where these defaults come from

If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.

> 

Reviewer 2: Can we add one more sentence here that the UI doesn't support pinned resources for the `"admin"` perspective, like:

```suggestion
Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEP validations.
```

### API Extensions

Extend the existing `operator.openshift.io/v1` / `Console CRD` by adding `spec.customization.perspectives.pinnedResources`

So the example yaml should look like this:

To set default pre-pinned resources

```yaml
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
  ...
spec:
  customization:
    perspectives:
    - id: 'dev'
      ...
      pinnedResources:
      - group: apps
        version: v1
        kind: Deployment

>       - group: core

Reviewer 3: Group core isn't correct, it should be `""`

Reviewer 2: As mentioned above, should we maybe use apiVersion and kind instead, or is this triple fine?

        version: v1
        kind: ConfigMap
    - ...
  ...
```

### Risks and Mitigations

> N/A

Reviewer 3: Risk: A user provides an invalid combination of API/Group/Version
Risk: After a cluster is upgraded, a previously valid combination of API/Group/Version is now invalid

Reviewer 2: @debsmita1 @JoelSpeed 

**Risk 1:** Invalid resources are ignored already today because of:
**Risk 2:** This could already happen today. The user can add a pinned resource and then delete a CRD or uninstall an operator. These resources are then not shown anymore in the sidebar.

But it makes sense to address both risks and mitigations here.

I see one more risk:

**Risk 3:** Currently the user preferences save the default pinned resources also if the user doesn't touch them. This means that the user settings always contain the info "Secret", "ConfigMap".

We should a) check if it's easy to fix that the defaults are not saved when the user doesn't touch them.

And b) Should we also use the cluster pinned resources when the pinned resources is (exactly) equal "Secret", "ConfigMap"?

### Drawbacks

N/A

## Design Details

> We will provide a configuration UI and a yaml snippet in the

Reviewer 2: ```suggestion
The list of available Kubernetes resources could be read via `kubectl api-resources`.

The console will also provide a configuration UI and a YAML snippet in the
```

console to provide a list of available resources that can be pinned to the navigation.

### Test Plan

- If no pinnedResources are defined, then default pinned resources should be visible to users who do not have their navigation items customized
- If pinnedResources is an empty list, then default pinned resources should be visible to users who do not have their navigation items customized
- If the pinnedResources list is non-empty, then users who do not have their navigation customized should see the pinned resources defined in the customization

### Graduation Criteria

> This will be released directly to GA.

Reviewer 3: There is no benefit to introducing this as a tech preview first?

Reviewer 2: I don't think this is required, but a good idea for customizations in general.

@debsmita1 we can check this with Serene when she is back from PTO.

Reviewer 3: There is a movement within openshift, to improve stability, that all new features should be introduced first as tech previews, and then to GA. I guess it's not mandatory but something to consider as we are releasing in a more continuous fashion

Author: @serenamarie125  ^

Author: Resolved in this conversation https://coreos.slack.com/archives/CHGU4P8UU/p1670435710137509

We can add `spec.customization.perspectives.pinnedResources` which lists default pre-pinned resources

### Workflow Description

To add default pre-pinned resources:

```text
├── spec
│   ├── customization
|       ├── perspectives
|           ├── id
|           ├── ...
|           ├── pinnedResources
|               └── group
                └── version
                └── resource
└── ...
```

Each perspective conforms to the following schema :

- `pinnedResources: configv1.GroupVersionResource[]` , defines a list of pinned resources. The default pinned resources are contributed by the console UI extension plugin.  
- `group: string` , defines the group of the resource.
- `version: string` , defines the version of the resource.
- `resource: string` , defines the type of the resource.

If the `pinnedResources` list is empty, the default pinned resources contributed by the console UI extension plugin will be shown.
If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.
If the `group`, `version` or `resource` is incorrect, the resource to be pinned will be ignored.

> Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEL validations.

Reviewer 2: my fault :)

```suggestion
Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEL validation.
```

![Default pinned resources](default-pinned-resources.png)

## Motivation

Some customers want to provide default pre-pinned resources for all users.

### User Stories

#### Story 1

Customer wants to customize the developer perspective navigation by providing pre-pinned resources to users who do not have their navigation items customized.

As a cluster admin, I want to provide default pre-pinned resources for all users so that the developers who donot have their navigation items customized will see these default pinned resources.

### Goals

- Allow cluster admins to define default pre-pinned resources on the dev perspective navigation

### Non-Goals

N/A

## Proposal

We can extend the existing `operator.openshift.io/v1` / `Console CRD`.
We can add `spec.customization.perspectives.pinnedResources` which lists default pre-pinned resources

### Workflow Description

To add default pre-pinned resources:

```text
├── spec
│   ├── customization
|       ├── perspectives
|           ├── id
|           ├── ...

> |           ├── pinnedResources

Reviewer 4: The `perspectives` stanza already exits and we're only adding `pinnedResources`, correct?

Author: That's right

|               └── group
                └── version
                └── resource
└── ...
```

Each perspective conforms to the following schema :

- `pinnedResources: operatorv1.PinnedResourceReference[]` , defines a list of pinned resources. The default pinned resources are contributed by the console UI extension plugin.  
- `group: string` , defines the group of the resource.
- `version: string` , defines the version of the resource.
- `resource: string` , defines the type of the resource. It is the plural form of the resource kind in lowercase.

If the `pinnedResources` list is empty, the default pinned resources contributed by the console UI extension plugin will be shown.
If the `pinnedResources` list is non-empty, only the listed items will be shown if the user hasn't already customized the navigation items.

> If the `group`, `version` or `resource` is incorrect, the resource to be pinned will be ignored.

Reviewer 4: What do you mean by "incorrect"? Do you mean the resource doesn't exist on the cluster?

Have we considered how this will work with multi-cluster in 4.13 where console talks to different clusters that might have different sets of resources?

Reviewer 2: Pinned resources aren't shown when the CRD (model) isn't found. So switching should not be an issue. 

And isn't the customization configuration per cluster? Is the index and Server flags reloaded when Switching a cluster?

We should def. test this.

Author: When switching between clusters the invalid resources should disappear. We will test this

Reviewer 4: The console customization comes from the hub since that is where the console is running. We don't read console configuration from the spoke clusters.

Reviewer 4: I would clarify in the enhancement what you mean by "incorrect" here.

> Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEL validation.

Reviewer 3: Has this been included in the API PR?

Author: Yes it has been included

> Note: The Administrator perspective (id="admin") doesn't support pinned resources at the moment and can be excluded via CEL validation.

Reviewer 4: What about unrecognized perspectives / perspectives contributed by plugins?

Author: So, should we update the CEL to allow adding pinnedResources only to the "dev" perspective?
cc @jerolimov

Reviewer 2: Other perspectives supports pinned resources as well. I think any other then admin should be fine. So we are good here I would expect at the moment.

Author: > Hm, I don't see a way to pin resources to the "All Clusters" perspective for ACM/MCE when I test it in the UI. We should double check this.
> 
I have updated this to:
```
Note: The Administrator perspective (id="admin") and "All Clusters" perspective for ACM/MCE (id="acm") doesn't support pinned resources at the moment and can be excluded via CEL validation.
```

> @jerolimov We should also test that we don't have a race condition here:
> 
> https://github.com/openshift/console/blob/master/frontend/packages/console-app/src/components/nav/PerspectiveNav.tsx#L28-L31
> 
> We perform API discovery asynchronously, and the cached resources won't always match the current on initial console load. Also on the first load there will be no cached resources, which could hide links to some resources when a user first uses the console. Really we should react when the models change.

We tested this by clearing the local storage and then refreshed the page and here's a small recording of that, please let us know if there's any other way to test this:

https://user-images.githubusercontent.com/22490998/207379291-5fa26aae-2815-4b43-93bb-c56f2384d7e7.mov

> Note: The Administrator perspective (id="admin") and "All Clusters" perspective for ACM/MCE (id="acm") doesn't support pinned resources at the moment and can be excluded via CEL validation.

Reviewer 4: I don't think other perspectives contributed by plugins will get pinned resources by default. We should clarify that here. Basically at the moment, we should only allow this for the dev perspective (unless I'm mistaken).

### API Extensions

Extend the existing `operator.openshift.io/v1` / `Console CRD` by adding `spec.customization.perspectives.pinnedResources`

So the example YAML should look like this:

To set default pre-pinned resources

```yaml
apiVersion: operator.openshift.io/v1
kind: Console
metadata:
  name: cluster
  ...
spec:
  customization:
    perspectives:
    - id: 'dev'
      ...
      pinnedResources:
      - group: apps
        version: v1

>         resource: Deployment

Reviewer 3: We need to fix this, resource is `deployments` not `Deployment`

      - group: ""
        version: v1
        resource: configmaps
    - ...
  ...
```

### Risks and Mitigations

**Risk 1:** A cluster admin configures an invalid combination of API/Group/Version

> Mitigation: Invalid combination of group, version and resource will not be shown to users. The cluster admin interface can show them with a warning.

Reviewer 4: Where is the warning displayed?

I'd probably only show the warning to admins who have permission to fix the problem.

Author: The warning is shown as an Alert on submission of the yaml

Author: we have added a regex to validate if the values entered for resource, version, and group are correct

Reviewer 2: We can show the warning just in the "console cluster configuration" UI, where the admin can configure it.

We now the available CRDs there and can show configured entries that will not shown to the user because there is no CRD/resource available.

Like this:

```yaml
perspectives:
  id: dev
  pinnedResources:
  - resource: Secret
  - resource: Pipelinee # wrong with two e
```

The UI could show:

- (S) Secret
- (P) Pipelinee :warning: with a tooltip "Missing CRD, this pinned resource might not be shown."

Reviewer 2: @lokanandaprabhu please take care of invalid configurations in the UI and show them anyway with such a warning.

Reviewer 4: This gets a bit tricky since we do API discovery in the background, but I wouldn't show a permanent warning to users who don't have a way to fix the problem. Normal users won't be able to edit this config. It would be better to hide the nav item IMO.

I guess there is the case where the user pins a custom resource themselves and then the CRD is removed, but we don't have a way to tell, correct?

Author: when the user pins a custom resource themselves and the CRD is removed, the pinned resource is no longer shown on the navigation.

After discussing this with @jerolimov, we would show a warning alert icon on the customization form for the unrecognized resource, on the navigation the user would only see the recognized ones.

Reviewer 4: > After discussing this with @jerolimov, we would show a warning alert icon on the customization form for the unrecognized resource, on the navigation the user would only see the recognized ones.

Sounds good to me.

We should clarify again what we mean by "incorrect" here. "Missing" might be a better term.

**Risk 2:** After a cluster is upgraded, a previously valid combination of API/Group/Version might get invalid
Mitigation: Invalid combination of group, version and resource will not be shown to users. The cluster admin interface can show them with a warning.

**Risk 3:** Currently, also if the user have not customized the pinned resources, the default settings are saved in user settings. After the update we can not determinate if the user touched them or not.

> Mitigation: For existing users where the saved pinned resources matches exactly the default pinned resources (from the perspective), the UI will show the customized default pinned resource (changed by the cluster admin).

Reviewer 4: Won't this only catch the first time an admin changes the pinned resources? For instance, if the admin adds deployments one week and config maps the next... The config maps won't get added since the previous user setting was not the default (even if the user didn't customize anything).

Reviewer 2: We have three buckets with pinned resources per perspective

1. Pinned resources from the perspective
2. Pinned resources from the customization
3. Pinned resources from the user settings

We should NOT save pinned resources until the user changes them manually (Add/Remove/Reorder). Unluckily we do this at the moment. So would recommend changing this, and then we need also to migrate the old data.

I hope we can do this while loading the data the first time like this:

1. If the user has pinned resources in the user settings:
    1. If these from the user settings are **_"deep" equal_** to the pinned resources from the perspective => delete the PR from user settings.
    2. Otherwise use the pinned resources from the user settings.
2. If there are pinned resources in the customization use them
3. Fallback to the pinned resources from the plugin definition.

We could also think about a more explicit migration, but I would keep it simple here and just compare both lists.

Reviewer 4: @jerolimov's proposal here looks reasonable to me. We should update the enhancement to cover this.
